[flutter_tools] Let CMake handle escaping

jpnurmi opened the pull request:
## Description

Using `VERBATIM` for [`add_custom_command()`](https://cmake.org/cmake/help/latest/command/add_custom_command.html) handles escaping of paths with spaces etc. correctly as long as the manual escaping in `flutter/ephemeral/generated_config.cmake` is removed. A migrator has been added to ensure that existing projects are updated accordingly.

## Related Issues

Fixes: #67270

## Tests

Minor updates:
- test/commands.shard/hermetic/build_linux_test.dart
- test/commands.shard/hermetic/build_windows_test.dart 

Added:
- test/general.shard/migrations/cmake_project_migration_test.dart

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

jonahwilliams commented:
Are there other characters that need to be un-escaped besides spaces?

jpnurmi commented:
I had a quick look. I think the full list is here: :exploding_head: 

https://github.com/Kitware/CMake/blob/master/Source/cmOutputConverter.cxx#L229-L277

jonahwilliams commented:
Thats not that bad!

jpnurmi commented:
Should any other of those variables be unescaped? Maybe BUNDLE_SKSL_PATH?

jonahwilliams commented:
Yeah, in theory anything that is a path string could require unescaping. 

Because this will require more logic, I would actually create the unescape function in a new file in lib/src (with no dependencies). Then it can still have a unit test

stuartmorgan commented:
Does adding `VERBATIM` [here](https://github.com/flutter/flutter/blob/145df3995379f5c1d3a613b249080a3b0c32a938/packages/flutter_tools/templates/app/linux.tmpl/flutter/CMakeLists.txt#L80) instead fix these? These are already quoted, so shouldn't need escaping; ideally we would just turn that behavior off.

jpnurmi commented:
With `VERBATIM`:

```
$ flutter build linux -v
[   +1 ms] cd "/home/jpnurmi/Temp/foo bar/foo_bar/build/linux/release/flutter" && /usr/bin/cmake -E env FLUTTER_ROOT=\"/home/jpnurmi/Flutter\" "PROJECT_DIR=\"/home/jpnurmi/Temp/foo bar/foo_bar\""
DART_OBFUSCATION=\"false\" TRACK_WIDGET_CREATION=\"true\" TREE_SHAKE_ICONS=\"true\" "PACKAGE_CONFIG=\"/home/jpnurmi/Temp/foo bar/foo_bar/.dart_tool/package_config.json\"" FLUTTER_TARGET=\"lib/main.dart\"
/home/jpnurmi/Flutter/packages/flutter_tools/bin/tool_backend.sh linux-x64 Release
[        ] /home/jpnurmi/Flutter/packages/flutter_tools/bin/tool_backend.sh: line 9: /home/jpnurmi/Temp/foo bar/foo_bar/build/linux/release/flutter/"/home/jpnurmi/Flutter"/bin/cache/dart-sdk/bin/dart: No such file or directory
```

For reference, without `VERBATIM`:

```
[   +1 ms] cd "/home/jpnurmi/Temp/foo bar/foo_bar/build/linux/release/flutter" && /usr/bin/cmake -E env FLUTTER_ROOT="/home/jpnurmi/Flutter" PROJECT_DIR="/home/jpnurmi/Temp/foo\ bar/foo_bar"
DART_OBFUSCATION="false" TRACK_WIDGET_CREATION="true" TREE_SHAKE_ICONS="true" PACKAGE_CONFIG="/home/jpnurmi/Temp/foo\ bar/foo_bar/.dart_tool/package_config.json" FLUTTER_TARGET="lib/main.dart"
/home/jpnurmi/Flutter/packages/flutter_tools/bin/tool_backend.sh linux-x64 Release
```

jpnurmi committed:
[flutter_tools] let CMake escape paths with special characters

Use of VERBATIM for add_custom_command() to make CMake take care of
escaping paths with spaces and other special characters on Windows and
Linux. Manual escaping in flutter/ephemeral/generated_config.cmake had
to be removed to avoid double escaping.

Includes a CMake project migrator to ensure that generated_config.cmake
in existing projects is updated accordingly.

Fixes: #67270

jpnurmi commented:
Using `VERBATIM` does the right thing as long as the manual escaping in `flutter/ephemeral/generated_config.cmake` is removed.

stuartmorgan commented:
Great, that sounds like the right solution then given that the docs recommend `VERBATIM` (I just didn't see that when putting that generation code together).

jonahwilliams commented on packages/flutter_tools/lib/src/cmake.dart:
```diff
@@ -40,12 +40,12 @@ file(TO_CMAKE_PATH "$escapedProjectDir" PROJECT_DIR)
 
 # Environment variables to pass to tool_backend.sh
 list(APPEND FLUTTER_TOOL_ENVIRONMENT
-  "FLUTTER_ROOT=\\"$escapedFlutterRoot\\""
```
Will changing these paths break existing projects that don't already have VERBATIM in their templates?

Also, you need to update the test expectation for this

jpnurmi replied:
> Will changing these paths break existing projects that don't already have VERBATIM in their templates?

I thought about the same first, but I saw that both `buildLinux()` and `buildWindows()` are unconditionally re-writing `generated_config.cmake`.

> Also, you need to update the test expectation for this

Thanks, done

stuartmorgan replied:
> Will changing these paths break existing projects that don't already have VERBATIM in their templates?

Ugh, good point. We'll need to either write a migrator, or make this logic conditional on the presence of `VERBATIM`.

> but I saw that both `buildLinux()` and `buildWindows()` are unconditionally re-writing `generated_config.cmake`.

Right, but the problem is that `CMakeLists.txt` is only created at `flutter create`-time, and that's where the VERBATIM needs to be added.

jpnurmi replied:
Right. What kind of migrator did you have in mind? I wonder if it's really worth the trouble. Perhaps going back to the earlier solution (unescaping `'\ '`) would be good enough? It fixes the most common case in a simple and backward-compatible way.

jonahwilliams replied:
The tool already has a migrator API used for iOS but not yet for linux.

See https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/ios/migrations/ios_migrator.dart and https://github.com/flutter/flutter/blob/f865ac7e259d320d7e565f887f894f293a463782/packages/flutter_tools/lib/src/ios/migrations/xcode_build_system_migration.dart#L13

jpnurmi commented:
WIP: missing Windows & tests

jpnurmi commented:
Alright, this is ready for another round of reviews.

Where do you prefer the have shared CMake migrators and tests for Linux & Windows?

jpnurmi commented:
Rebased. https://github.com/flutter/flutter/pull/71757 applied similar migrator changes to make it available on all platforms.

jmagman commented:
> Rebased. #71757 applied similar migrator changes to make it available on all platforms.

Thanks for noticing this, I meant to comment here.  The migration part LGTM.

@jonahwilliams is OOO.  @stuartmorgan does the actual `VERBATIM` part of this PR look good?  Can I merge it?

stuartmorgan commented on packages/flutter_tools/test/commands.shard/hermetic/build_linux_test.dart:
```diff
@@ -362,17 +362,17 @@ clang: error: linker command failed with exit code 1 (use -v to see invocation)
+      '  "FLUTTER_ROOT=$_kTestFlutterRoot"',
+      '  "PROJECT_DIR=${fileSystem.currentDirectory.path}"',
+      r'  "FLUTTER_TARGET=lib/other.dart"',
+      r'  "BUNDLE_SKSL_PATH=foo/bar.sksl.json"',
```
Nit: you should be able to drop all the `r` prefixes here; they were to simplify the quote escaping.

stuartmorgan-g reviewed (approved):
Sorry, I too was OOO. The actual CMake changes here LGTM; someone else should review the migrator code, then it's good to land.

jpnurmi committed:
[flutter_tools] drop raw string prefixes in build_linux_test.dart
